[E2E] Introduce SPIR-V Backend support in SYCL tests and end-to-end tests

VyacheslavLevytskyy opened the pull request:
Introduce SPIR-V Backend support in SYCL tests and end-to-end tests by adding XFAIL & UNSUPPORTED to tests that are not passed at the moment when Khronos Translator is substituted by SPIR-V Backend. Three syntax-only tests are moved from e2e to tests.

VyacheslavLevytskyy commented:
@AlexeySachkov @sarnex Please have a look. This PR shows how we possibly can enable the LLVM SPIR-V Backend in DPC++ CI. A separate question is how to pass "-fsycl-use-spirv-backend-for-spirv-gen" in the SYCL CTS test suite.

AlexeySachkov commented on .github/workflows/sycl-nightly.yml:
```diff
@@ -102,6 +102,48 @@ jobs:
+            tests_selector: e2e
+            extra_lit_opts: --param spirv-backend=True --param gpu-intel-gen12=True
+
+          - name: SYCL-CTS on OCL CPU w/ LLVM SPIR-V Backend
+            runner: '["Linux", "gen12"]'
+            image: ghcr.io/intel/llvm/ubuntu2204_intel_drivers:latest
+            image_options: -u 1001 --device=/dev/dri --privileged --cap-add SYS_ADMIN
+            target_devices: opencl:cpu
+            tests_selector: cts
+            extra_lit_opts: --param spirv-backend=True --param gpu-intel-gen12=True
+
+          - name: SYCL-CTS on L0 gen12 w/ LLVM SPIR-V Backend
+            runner: '["Linux", "gen12"]'
+            image: ghcr.io/intel/llvm/ubuntu2204_intel_drivers:latest
+            image_options: -u 1001 --device=/dev/dri --privileged --cap-add SYS_ADMIN
+            target_devices: level_zero:gpu
+            tests_selector: cts
+            extra_lit_opts: --param spirv-backend=True --param gpu-intel-gen12=True
```
I have no objections to that conceptually, but I have a concern that we may not have enough HW available to finish that testing in reasonable time,

VyacheslavLevytskyy replied:
Yes, I agree. I will keep the PR draft until we have completed discussing changes and potential consequences.

sarnex replied:
yeah especially since we're using gen12 which takes forever to run as it is. can you bring this up at our next gatekeeper/CI meeting? there is one today but idk if there is time for this issue. will fwd.

VyacheslavLevytskyy replied:
Sure, thank you

AlexeySachkov reviewed (approved):
`sycl/test-e2e` changes LGTM

sarnex reviewed (changes requested):
needs discussion at CI meeting imo

sarnex commented:
just E2E should be fine IMO, it runs in the middle of the night when nobody is working and CI is empty

AlexeySachkov commented on sycl/test-e2e/Basic/backend_info.cpp:
```diff
@@ -1,7 +1,7 @@
 // RUN: %{run} %t.out
 //
-// RUN: %{build} -DTEST_ERRORS -D_GLIBCXX_USE_CXX11_ABI=0 -fsyntax-only -Xclang -verify -Xclang -verify-ignore-unexpected=note
+// RUN: %{build} -DTEST_ERRORS -D_GLIBCXX_USE_CXX11_ABI=0 -fsyntax-only -Wno-error=unused-command-line-argument -Xclang -verify -Xclang -verify-ignore-unexpected=note
```
Ideally, `-fsyntax-only` check should be moved into `check-sycl`. That's not a strong comment in this case, because it is a "combined" test (i.e. syntax + execution), but still

AlexeySachkov commented on sycl/test-e2e/Basic/fpga_tests/fpga_io_pipes.cpp:
```diff
@@ -1,4 +1,4 @@
-// RUN: %{build} -fsyntax-only -o %t.out
+// RUN: %{build} -fsyntax-only -Wno-error=unused-command-line-argument -o %t.out
```
If we don't execute a test, then we should move it to `sycl/test`, i.e. `check-sycl`. From there we can decide if we want to also have a copy of it which we execute.

In this case, I object against the change and I would happily approve move of the test (except that I'm not a code owner :)). Feel free to send a separate PR, we should be able to merge it almost instantly

VyacheslavLevytskyy replied:
The test case is moved to sycl/test.

AlexeySachkov commented on sycl/test-e2e/NonUniformGroups/is_fixed_topology.cpp:
```diff
@@ -1,4 +1,4 @@
-// RUN: %{build} -fsyntax-only -o %t.out
+// RUN: %{build} -fsyntax-only -Wno-error=unused-command-line-argument -o %t.out
```
Once again, let's move the test instead of modifying it. This one I'm even aware of and have it my backlog, but who knows when I get to it

sarnex commented on sycl/test-e2e/format.py:
```diff
@@ -175,13 +175,21 @@ def execute(self, test, litConfig):
+                + fsycl_use_spirv_backend,
             )
         )
         if platform.system() == "Windows":
```
can we append to `substitutions` like we do here instead of adding to the main substituion step?

sarnex commented on sycl/test-e2e/format.py:
```diff
@@ -175,13 +175,21 @@ def execute(self, test, litConfig):
                 "%{build}",
-                "%clangxx -fsycl -fsycl-targets=%{sycl_triple} %verbose_print %s",
+                "%clangxx -fsycl -fsycl-targets=%{sycl_triple} %verbose_print %s"
+                + fsycl_use_spirv_backend,
```
nit: i dont think we need `fsycl` since this is just a local var name and `sycl-...` makes more sense to me

sarnex reviewed (commented):
initial review!

aelovikov-intel commented on .github/workflows/sycl-linux-precommit.yml:
```diff
@@ -127,6 +127,24 @@ jobs:
+            image: ghcr.io/intel/llvm/ubuntu2404_intel_drivers:latest
+            image_options: -u 1001 --device=/dev/dri --privileged --cap-add SYS_ADMIN
+            target_devices: opencl:cpu
+            extra_lit_opts: -j 50 --param spirv-backend=True
```
This makes no sense, these runners have desktop CPUs and you'll oversubscribe them.

aelovikov-intel commented on .github/workflows/sycl-nightly.yml:
```diff
@@ -87,6 +87,32 @@ jobs:
+            image_options: -u 1001 --privileged --cap-add SYS_ADMIN
+            target_devices: opencl:cpu
+            tests_selector: e2e
+            extra_lit_opts: --param spirv-backend=True --param gpu-intel-gen12=True
```
`gpu-intel-gen12` isn't required after https://github.com/intel/llvm/pull/16220.

aelovikov-intel reviewed (commented):
I think this should be split into several patches:

1) sycl-linux-run-tests.yml so that the testing can be dispatched manually.
2) Update tests and verify via manual CI runs
3) CI enabling in pre-commit/nightly. Shouldn't include many of the tests changes at that poing.

I also see some supposedly functional changes, they should go into separate PR as well.

YuriPlyakhin reviewed (approved):
Joint Matrix change LGTM

dkhaldi reviewed (approved):
JM changes LGTM

VyacheslavLevytskyy commented:
It was agreed that we split https://github.com/intel/llvm/pull/16317 into three parts:
- CI
- Clang driver
- XFAIL & UNSUPPORTED in end-to-end tests.

I reverted changes in CI/Clang here, so that this PR is only about introducing XFAIL & UNSUPPORTED into end-to-end tests now.

@sarnex Can you please have a look and unblock if everything is ok

aelovikov-intel commented:
> @sarnex Can you please have a look and unblock if everything is ok

IMO, this PR shouldn't even be outside draft state until https://github.com/intel/llvm/pull/16317 is merged in.

VyacheslavLevytskyy commented:
@sarnex @aelovikov-intel We have https://github.com/intel/llvm/pull/16600 and https://github.com/intel/llvm/pull/16602 merged. This PR contains only changes in test cases, and we got them approved as well. Are we good to proceed with a merge of this PR?

sarnex commented on sycl/test-e2e/InvokeSimd/lit.local.cfg:
```diff
@@ -6,3 +6,6 @@ config.required_features += ['gpu']
 if platform.system() != "Linux":
    config.unsupported = True
+
+# At the moment SPIR-V Backend has no plans to suport this feature.
```
```suggestion
# At the moment SPIR-V Backend has no plans to support this feature.
```

sarnex commented on sycl/test-e2e/ESIMD/lit.local.cfg:
```diff
@@ -21,3 +21,6 @@ for substitution in config.substitutions:
 config.substitutions.insert(0,
   ("%clangxx", original_clangxx+" -Wno-error=deprecated-declarations"))
+
+# At the moment SPIR-V Backend has no plans to suport this feature.
```
```suggestion
# At the moment SPIR-V Backend has no plans to support this feature.
```

sarnex reviewed (approved):
lgtm just typos

aelovikov-intel commented:
> @sarnex @aelovikov-intel We have #16600 and #16602 merged. This PR contains only changes in test cases, and we got them approved as well. Are we good to proceed with a merge of this PR?

I'd like to see some CI results before doing that. Are we in a state when that can be done with manual dispatch of "SYCL E2E" task?

aelovikov-intel commented:
origin/sycl: https://github.com/intel/llvm/actions/runs/12814470669
this PR merged with origin/sycl: https://github.com/intel/llvm/actions/runs/12815172832
581 failures -> 575 failures.

It doesn't look like it's working properly. @VyacheslavLevytskyy can you take a look into this?

aelovikov-intel reviewed (approved):
Before this PR: https://github.com/intel/llvm/actions/runs/12814470669 (581 fails)
With this PR: https://github.com/intel/llvm/actions/runs/12815598260 (4 fails)

VyacheslavLevytskyy commented:
> Before this PR: https://github.com/intel/llvm/actions/runs/12814470669 (581 fails) With this PR: https://github.com/intel/llvm/actions/runs/12815598260 (4 fails)

Thank you. I will update those prior to CI is updated.
